src/org/jitsi/impl/neomedia/transform/fec/AbstractFECReceiver.java

```diff
@@ -0,0 +1,280 @@
+     * can be optimized. We very rarely (when we receive a packet out of order)
+     * need to insert at an arbitrary location.
+     */
+    protected final SortedMap<Integer, RawPacket> mediaPackets
+        = new TreeMap<Integer, RawPacket>(seqNumComparator);
+
+    /**
+     * Buffer which keeps (copies of) received fec packets.
+     *
+     * We keep them ordered by their RTP sequence numbers, so that
+     * we can easily select the oldest one to discard when the buffer is
+     * full (when the map has more than <tt>FEC_BUFF_SIZE</tt> entries.
+     *
+     * We keep them in a <tt>Map</tt> so that we can easily search for a
+     * packet with a specific sequence number.
+     *
+     * Note: This might turn out to be inefficient, especially with increased
+     * buffer sizes. In the vast majority of cases (e.g. on every received
+     * packet) we do an insert at one end and a delete from the other -- this
+     * can be optimized. We very rarely (when we receive a packet out of order)
+     * need to insert at an arbitrary location.
+     */
+    protected final SortedMap<Integer,RawPacket> fecPackets
+        = new TreeMap<Integer, RawPacket>(seqNumComparator);
+
+
+    /**
+     * Initialize the FEC receiver
+     * @param ssrc the ssrc of the stream on which fec packets will be received
+     * @param payloadType the payload type of the fec packets
+     */
+    AbstractFECReceiver(long ssrc, byte payloadType)
+    {
+        this.ssrc = ssrc;
+        this.payloadType = payloadType;
+    }
+
+    /**
+     * Saves <tt>p</tt> into <tt>fecPackets</tt>. If the size of
+     * <tt>fecPackets</tt> has reached <tt>FEC_BUFF_SIZE</tt> discards the
+     * oldest packet from it.
+     * @param p the packet to save.
+     */
+    private void saveFec(RawPacket p)
+    {
+        if (fecPackets.size() >= FEC_BUF_SIZE)
+            fecPackets.remove(fecPackets.firstKey());
+
+        fecPackets.put(p.getSequenceNumber(), p);
+    }
+
+    /**
+     * Makes a copy of <tt>p</tt> into <tt>mediaPackets</tt>. If the size of
+     * <tt>mediaPackets</tt> has reached <tt>MEDIA_BUFF_SIZE</tt> discards
+     * the oldest packet from it and reuses it.
+     * @param p the packet to copy.
+     */
+    protected void saveMedia(RawPacket p)
+    {
+        RawPacket newMedia;
+        if (mediaPackets.size() < MEDIA_BUF_SIZE)
+        {
+            newMedia = new RawPacket();
+            newMedia.setBuffer(new byte[FECTransformEngine.INITIAL_BUFFER_SIZE]);
+            newMedia.setOffset(0);
+        }
+        else
+        {
+            newMedia = mediaPackets.remove(mediaPackets.firstKey());
+        }
+
+        int pLen = p.getLength();
+        if (pLen > newMedia.getBuffer().length)
+        {
+            newMedia.setBuffer(new byte[pLen]);
+        }
+
+        System.arraycopy(p.getBuffer(), p.getOffset(), newMedia.getBuffer(),
+            0, pLen);
+        newMedia.setLength(pLen);
```

Comment:
I suppose that these packets always have offset 0 (and I realize this might be existing code being moved), but I think an explicit setOffset(0) would be good to have.

Reply:
done